### ISSUE TITLE
Error message while trying to read non-existing file

### ISSUE THREAD

**Issue by alex-konovalov** [gap-system/gap#354]
@cdwensley suggested that the following error message is not very helpful - one would expect a "file does not exist" - in the following attempt:
```
gap> Read(Filename(DirectoriesLibrary("tst"),"testall.g"));
Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 1st choice method found for `Read' on 1 arguments called from
<function "HANDLE_METHOD_NOT_FOUND">( <arguments> )
called from read-eval loop at line 1 of *stdin*
you can 'quit;' to quit to outer loop, or
you can 'return;' to continue
```
Of course (this is in GAP 4.8.0), `Filename` returned `fail` because `testall.g` is now `teststandard.g`. We can't detect this in `Read` since it is an operation declared for strings, and installed for streams via `InstallOtherMethod`. We need either to declare it for IsObject to catch `fail`, or install other method for booleans, like e.g. this:

```
InstallOtherMethod( Read, "to produce meaningful error message",
[ IsBool ],
function( x )
if x = fail then
    Error("Calling Read(fail) is not possible. Please check that the file exists");
else
    Error("The argument of 'Read' should be a stream, or a string with the name of the existing file");
fi;
end);
```

Does it make sense to add such other method?

**Comment by ChrisJefferson** [comment]
Another option which might make what went on easier would be to print out the arguments.

You can get the arguments in that break loop with 'ShowArguments', but I bet not many people know that, and it seems (to me) just printing the arguments out might be a good idea, rather than just '1 arguments'.

However, at some point someone clearly decided that was a bad idea, as they implemented 'ShowArguments'. I wonder why...

**Comment by ChrisJefferson** [comment]
For example, given a (partial) patch, using `PrintString`, the error becomes:

```
Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 1st choice method found for Read(fail) called from
<function "HANDLE_METHOD_NOT_FOUND">( <arguments> )
 called from read-eval loop at line 1 of *stdin*
you can 'quit;' to quit to outer loop, or
you can 'return;' to continue
```

**Comment by alex-konovalov** [comment]
But imagine if method fails on a 1000x1000 matrix and it will be printed in this message... That was perhaps the reason behind adding the functionality of `ShowArguments`, so the initial error message is compact.

**Comment by ChrisJefferson** [comment]
We could print out a truncated variable, something like (passing a big matrix to Read)

```
... method found for Read([[1,2,3,4,... ) called from
```

**Comment by alex-konovalov** [comment]
Ah, that makes things better.  Such change may help for no-method-found errors to cause less confusion and take less time to understand. In theory, for any object there a fallback method should print `<object>` in case no specific methods are installed, so this will not cause another no-method-found-error. Can we look at your patch?

**Comment by ChrisJefferson** [comment]
Unfortunately this patch has proved to be a pain, because it's hard to turn an object into a string this early in library loading.

A basic proof of principle is at #355

**Comment by frankluebeck** [comment]
> @cdwensley suggested that the following error message is not very helpful - one would expect a "file does not exist" - in the following attempt:
```
gap> Read(Filename(DirectoriesLibrary("tst"),"testall.g"));
Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 1st choice method found for `Read' on 1 arguments called from
<function "HANDLE_METHOD_NOT_FOUND">( <arguments> )
called from read-eval loop at line 1 of *stdin*
you can 'quit;' to quit to outer loop, or
you can 'return;' to continue
```

I do not agree that a "file does not exist" error message is expected here. As documented `Filename(DirectoriesLibrary("tst"),"testall.g")` returns the name of an existing file or `fail`. It is a user error to feed its output into `Read` without checking the return value. 

Furthermore, I find the error message not so bad. For less experienced users it even gives the hint to read `?Recovery` to investigate method not found errors. It also says that the error is in the argument of `Read`. Even without calling `ShowDetails();` and `ShowArguments();` I find it easy to guess what went wrong here.

Of course, there is a good reason why the arguments are NOT printed in this error message (but only on demand via `ShowArguments()`: some GAP object are just so big that nobody ever wants to see them printed. Even if you only want to print the beginning by first computing the `String` of each argument this can take considerable time and memory. It is better to enter the break loop quickly than to start potentially extremely long computations.

I prefer to leave this as it is. 

@alex-konovalov: I would not install methods whose only purpose is to overwrite default error messages. If this was done extensively this would slow down method selection.

**Comment by ChrisJefferson** [comment]
@frankluebeck : I'll admit I've never called `?Recovery` before yesterday, and I suspect many people don't (we should). The message is a bit vague -- I assumed `debugging` meant I might want to fix some library problem (I just want to know what I did wrong). I didn't even know about `ShowDetails` or `ShowArguments`. Perhaps adding a second line to the error that says the following would help guide users more directly.

 `Error, ShowArguments() will show arguments, ShowDetails() gives other information`

**Comment by stevelinton** [comment]
I wonder if there is scope for an extra hint in the "No Method Found" message in the case where one of the arguments is `fail`. This often (for me usually) implies that the real problem was earlier  when I tried to open a non-existent file, or invert a singular matrix or .... Something like "3rd argument is 'fail' which might indicate that the real problem was earlier"

As we slowly move to return `fail`, rather than giving an error in more and more such situations, which is usually right as it allows the programmer to handle the problem if they choose to, the penalty is that error messages get less and less useful and this would be a small step to help with that.

**Comment by alex-konovalov** [comment]
@frankluebeck I had performance in mind while suggesting this, but my thought was that since this is concerning the `Read` operation, performance is slow anyway, so adding an extra method would not cause a big penalty.

@ChrisJefferson: agree that adding `Error, ShowArguments() will show arguments, ShowDetails() gives other information` will be more user friendly. You would have a chance of knowing them earlier in this case.

@stevelinton is it guaranteed that `fail` equals to `fail` and for any other object the method to return false for `fail=obj` always work? Can't find this - perhaps it is in the kernel. If so, looks like this idea should work.

I suggest to replace my PR with better error messages then, incorporating two suggestions by Steve and Chris.

**Comment by stevelinton** [comment]
@alex-konovalov Yes. There is just one fail and it will reliably test unequal to anything else.

**Comment by alex-konovalov** [comment]
@ChrisJefferson would you be willing to change your #355 and replace printing abbreviated arguments by a warning if one of the arguments is `fail`, or should I try to do this in a new PR?

**Comment by alex-konovalov** [comment]
If #460 will be merged, then it would be fine to close this issue.

**Comment by alex-konovalov** [comment]
Implemented in #460.
